rclcpp/include/rclcpp/executor.hpp

```diff
@@ -175,6 +175,26 @@ class Executor
   virtual void
   spin_some(std::chrono::nanoseconds max_duration = std::chrono::nanoseconds(0));
 
+  /// Complete all available queued work without blocking.
+  /**
+   * This function can be overridden. The default implementation is suitable for a
+   * single-threaded model of execution.
+   * Adding subscriptions, timers, services, etc. with blocking callbacks will cause this function
+   * to block (which may have unintended consequences).
+   *
+   * To ensure all available queued work is executed, this function has to check if there is
+   * available work repeatedly.
+   * Thus, if the time that waitables take to be executed is longer than the period on which new waitables
+   * become ready, this will execute work repeatedly until `max_duration` has elapsed.
```

Comment:
If I'm reading the logic correctly, the description of this logic isn't quite correct (nor is `spin_some` documentation).

For `spin_some`, the executor will collect work *once*, and then execute all of the work that was available at that time until the timeout expires.

For `spin_all`, the executor will collect work repeatedly, and then execute all of the work that was available (and that comes in) until the timeout expires.

In both cases, once it has executed one callback, if there is no further work to do, it will return immediately.

Does that sound like the correct logic to you?  If so, I'd recommend updating this documentation describing more of that.

Reply:
That description is correct, I will try to update the doc blocks to make the difference clearer.

Reply:
> the executor will collect work repeatedly, and then execute all of the work that was available

To be more precise, `spin_all` collects and executes work repeatedly within the specified duration.

Reply:
See https://github.com/ros2/rclcpp/pull/1156/commits/07c84d36a8e665bc8d4c4d90df869b96d62cc012